src/Oracle/Oracle.cs

```diff
@@ -0,0 +1,377 @@
+                    }
+                }).Start();
+            }
+        }
+
+        public void ProcessRequest(StoreView snapshot, ulong requestId, OracleRequest request)
+        {
+            Log($"Process oracle request: {requestId}, requestTx={request.OriginalTxid}");
+
+            Uri.TryCreate(request.Url, UriKind.Absolute, out var uri);
+            OracleResponse response;
+            switch (uri.Scheme.ToLowerInvariant())
+            {
+                case "http":
+                case "https":
+                    try
+                    {
+                        var data = Https.Request(requestId, request.Url, request.Filter);
+                        response = new OracleResponse() { Id = requestId, Code = OracleResponseCode.Success, Result = data };
+                    }
+                    catch (Exception e)
+                    {
+                        Log($"Request error {e.Message}");
+                        response = new OracleResponse() { Id = requestId, Code = OracleResponseCode.Error, Result = Array.Empty<byte>() };
+                    }
+                    break;
+                default:
+                    Log($"{uri.Scheme.ToLowerInvariant()} is not supported");
+                    response = new OracleResponse() { Id = requestId, Code = OracleResponseCode.Forbidden, Result = Array.Empty<byte>() };
+                    break;
+            }
+
+            var responseTx = CreateResponseTx(snapshot, response);
+
+            Log($"Builded response tx:{responseTx.Hash} requestTx:{request.OriginalTxid} requestId: {requestId}");
+
+            ECPoint[] oraclePublicKeys = NativeContract.Oracle.GetOracleNodes(snapshot);
+            foreach (var account in Wallet.GetAccounts())
+            {
+                var oraclePub = account.GetKey().PublicKey;
+                if (!account.HasKey || account.Lock || !oraclePublicKeys.Contains(oraclePub)) continue;
+
+                var txSign = responseTx.Sign(account.GetKey());
+                AddResponseTxSign(snapshot, requestId, txSign, oraclePub, responseTx);
+                SendResponseSignature(requestId, txSign, account.GetKey());
+
+                Log($"Send oracle sign data: Oracle node: {oraclePub} RequestTx: {request.OriginalTxid} Sign: {txSign.ToHexString()}");
+            }
+        }
+
+        private Transaction CreateResponseTx(StoreView snapshot, OracleResponse response)
+        {
+            var oracleNodes = NativeContract.Oracle.GetOracleNodes(snapshot);
+            var m = oracleNodes.Length - (oracleNodes.Length - 1) / 3;
+            var n = oracleNodes.Length;
+            var oracleSignContract = Contract.CreateMultiSigContract(m, oracleNodes);
+
+            var tx = new Transaction()
+            {
+                Version = 0,
+                ValidUntilBlock = snapshot.Height + Transaction.MaxValidUntilBlockIncrement,
```

Comment:
I think, it should be calculated based on height of the request tx, not snapshot height.
This way tx can be created based solely on request and we don't rely on some constantly-changing state when signing tx.

Reply:
Done